Beta versions of the compiler support could be improved

With the following workflow file: https://github.com/ocaml/merlin/blob/56e0601ad40995421ae9e3537158af30559fb65f/.github/workflows/main.yml
```
 ocaml-compiler:
          - 4.13.0~beta1
```
Is not understood as a valid compiler version and will instead expect it to be a package name, which makes opam fail with:
```
> Install OCaml
  /opt/hostedtoolcache/opam/2.0.9/x86_64/opam switch create . --no-install --packages 4.13.0~beta1
  opam: option `--packages': invalid element in list (`4.13.0~beta1'): Invalid
        character in package name "4.13.0~beta1"
  am switch [OPTION]... [COMMAND] [ARG]...
  Try `opam switch --help' or `opam --help' for more information.
  Error: The process '/opt/hostedtoolcache/opam/2.0.9/x86_64/opam' failed with exit code 2
```


This is a valid version string for opam and expliciting `ocaml-base-compiler.4.13.0~beta1` shouldn't be required.

And using `ocaml-base-compiler.4.13.0~beta1` the result is:
```
   /opt/hostedtoolcache/opam/2.0.9/x86_64/opam switch create . --no-install --packages ocaml-base-compiler.4.13.0~beta1
  [ERROR] Could not resolve set of base packages:
          The following dependencies couldn't be met:
            - ocaml-base-compiler → ocaml-beta
                unmet availability conditions: enable-ocaml-beta-repository
                ```

--- a/packages/setup-ocaml/src/version.ts
+++ b/packages/setup-ocaml/src/version.ts
@@ -13,46 +13,63 @@ async function retrieveAllCompilerVersions() {
     repo: "opam-repository",
     path: "packages/ocaml-base-compiler",
   });
-  const versions = new Set<string>();
+  const versions = new Map<string, string>();
   if (Array.isArray(packages)) {
     for (const { path: p } of packages) {
       const basename = path.basename(p);
-      const version = basename.replace("ocaml-base-compiler.", "");
-      const parsed = semver.parse(version, { loose: true });
+      const opamVersion = basename.replace("ocaml-base-compiler.", "");
+      const parsed = semver.parse(opamVersion.replace("~", "-"), {
+        loose: true,
+      });
       if (parsed !== null) {
-        const { major, minor: _minor, patch } = parsed;
         const minor =
-          major < 5 && _minor < 10
+          parsed.major < 5 && parsed.minor < 10
             ? // ocaml-base-compiler.4.00.0, ocaml-base-compiler.4.01.0
-              `0${_minor}`
+              `0${parsed.minor}`
             : // ocaml-base-compiler.5.4.0, ocaml-base-compiler.4.14.2
-              _minor;
-        const version = `${major}.${minor}.${patch}`;
-        versions.add(version);
+              parsed.minor;
+        const prerelease =
+          parsed.prerelease.length > 0 ? `-${parsed.prerelease.join(".")}` : "";
+        const semverVersion = `${parsed.major}.${minor}.${parsed.patch}${prerelease}`;
+        versions.set(semverVersion, opamVersion);
       }
     }
   }
-  return [...versions];
+  return versions;
 }
 
 async function resolveVersion(semverVersion: string) {
-  const compilerVersions = await retrieveAllCompilerVersions();
-  const matchedFullCompilerVersion = semver.maxSatisfying(
-    compilerVersions,
-    semverVersion,
-    { loose: true },
-  );
-  if (matchedFullCompilerVersion === null) {
-    throw new Error(
-      `Could not find any OCaml compiler version matching '${semverVersion}' in the opam-repository. Please check if you specified a valid version number or version range.`,
-    );
+  const versions = await retrieveAllCompilerVersions();
+  const semverVersions = versions.keys().toArray();
+  const stableMatch = semver.maxSatisfying(semverVersions, semverVersion, {
+    loose: true,
+  });
+  if (stableMatch !== null) {
+    const opamVersion = versions.get(stableMatch);
+    if (opamVersion !== undefined) {
+      return opamVersion;
+    }
   }
-  return matchedFullCompilerVersion;
+  const prereleaseMatch = semver.maxSatisfying(semverVersions, semverVersion, {
+    loose: true,
+    includePrerelease: true,
+  });
+  if (prereleaseMatch !== null) {
+    const opamVersion = versions.get(prereleaseMatch);
+    if (opamVersion !== undefined) {
+      return opamVersion;
+    }
+  }
+  throw new Error(
+    `Could not find any OCaml compiler version matching '${semverVersion}' in the opam-repository. Please check if you specified a valid version number or version range.`,
+  );
 }
 
 export const resolvedCompiler = (async () => {
-  const resolvedCompiler = isSemverValidRange(OCAML_COMPILER)
-    ? `ocaml-base-compiler.${await resolveVersion(OCAML_COMPILER)}`
-    : OCAML_COMPILER;
-  return resolvedCompiler;
+  const semverInput = OCAML_COMPILER.replace("~", "-");
+  if (isSemverValidRange(semverInput)) {
+    const opamVersion = await resolveVersion(semverInput);
+    return `ocaml-base-compiler.${opamVersion}`;
+  }
+  return OCAML_COMPILER;
 })();

